extmod/asyncio/stream.py

```diff
@@ -30,6 +36,15 @@ async def wait_closed(self):
     def read(self, n=-1):
         r = b""
         while True:
+            if hasattr(self.s, "ssl_pending"):
```

Comment:
Is this `ssl_pending` only needed for the unix port to work correctly?  If so I would rather fix unix by making it work correctly when polling an ssl socket.

Reply:
I can confirm that the `tests/multi_net/asyncio_tls_server_client_readline.py` test works with the `ssl_pending` code there and fails when it is removed.

But, by enabling `MICROPY_PY_SELECT`, it will work without the `ssl_pending` code.

So I will fix the unix port so that it can work without needing to enable that option.

BTW, to run this test I actually had to regenerate the certificate and key, because otherwise it gave a "certificate has expired" error.

Reply:
> But, by enabling MICROPY_PY_SELECT, it will work without the ssl_pending code.

well that seems like an easy fix 😅

> BTW, to run this test I actually had to regenerate the certificate and key, because otherwise it gave a "certificate has expired" error.

I will update the tests too with a longer expiration date then 👍🏼

Reply:
> But, by enabling MICROPY_PY_SELECT, it will work without the ssl_pending code.
So I will fix the unix port so that it can work without needing to enable that option

Will wait for the fix then... 👀

Reply:
#12138 is now merged, which should fix the unix port so it doesn't need special handling for ssl-wrapped sockets.

Reply:
I found the issue with the readline test failing here, and #12195 should fix that.

Reply:
The above fix is now in master, so if you could please rebase this PR and see if the CI passes.

Also can you please re-test asyncio-mqtt.simple and asyncio-microdot.  If they don't work then please provide details how to reproduce the error.  Thanks!

Reply:
> Also can you please re-test asyncio-mqtt.simple and asyncio-microdot. If they don't work then please provide details how to reproduce the error. Thanks!

It's working again now 😮‍💨👍🏼 ! It must be #12195, so really nice work figuring this out  👏🏼


> The above fix is now in master, so if you could please rebase this PR and see if the CI passes

Done! and CI tests are passing now too.

Reply:
Very good, thanks for testing!

Reply:
After some further testing I've found another weird bug in unix port 👀, this time is the combination of a stream ssl socket and a plain stream socket .i.e running two `asyncio-microdot` web servers instances,  one with HTTP and the other with HTTPS. What happens is that at first both work BUT the moment I do a request to the HTTP instance,  then it will block the requests to the HTTPS instance until I do another request to the HTTP instance. 

I've also tested this with .e.g TASK A -> `async_urequests.py` (HTTP request) and TASK B ->`asyncio-mqtt.simple` (SSL) with the same results, i.e. `mqtt` client stops reading (writing still works) after the TASK A does an HTTP request, and then it will only work after TASK A does another HTTP request and so on.

The combination of HTTP-HTTP and HTTPS-HTTPS is OK. 
esp32 port is OK
Probably there is another case/flag missing like in #12195 which I'm not aware of 🤷🏼‍♂️

Reply:
I've updated `tests/net_inet` with `asyncio_http_https_opc.py` and `asyncio_https_http_opc.py` that reflects this behaviour, .i.e first test gets stuck after http request, second test works as expected. 
```
$ micropython asyncio_https_http_opc.py
write GET
read response
read: b'HTTP/1.1 200 OK'
close
done
write GET
read response
read: b'HTTP/1.1 200 OK'
close
done
$ micropython asyncio_http_https_opc.py
write GET
read response
read: b'HTTP/1.1 200 OK'
close
done
^CTraceback (most recent call last):
  File "asyncio_http_https_opc.py", line 109, in <module>
  File "asyncio/core.py", line 1, in run
  File "asyncio/core.py", line 1, in run_until_complete
  File "asyncio/core.py", line 1, in wait_io_event
KeyboardInterrupt:

```

Reply:
I've added `tests/multi_net/asyncio_tcp_tls_server_client.py` and `tests/multi_net/asyncio_tls_tcp_server_client.py` 

These tests cover the above edge case and as you can see `tests/multi_net/asyncio_tcp_tls_server_client.py` is failing...

Reply:
I think I've found what's going on with the HTTP request, it turns out that the function `poll_set_poll_once` https://github.com/micropython/micropython/blob/24a6e951ec7696b8d18d95fe5da36f7e489913d0/extmod/modselect.c#L252 in `extmod/modselect.c` gets called in an infinite loop after the first HTTP request. Doing an HTTPS request stops that loop but it also gets stuck until another HTTP request which starts the infinite loop again. e.g.

```
# HTTPS REQUEST
POLL
POLL
Request line
IOCTL REQUEST: 10
IOCTL REQUEST UNSUPPORTED: 10
NO FILE DESCRIPTOR
POLL
POLL
POLL
POLL
IOCTL REQUEST: 3
IOCTL REQUEST TO SOCKET
IOCTL REQUEST SOCK: 3
GET /temp HTTP/1.1
IOCTL REQUEST: 10
IOCTL REQUEST UNSUPPORTED: 10
NO FILE DESCRIPTOR
POLL
POLL
POLL
POLL
IOCTL REQUEST: 3
IOCTL REQUEST: 10
IOCTL REQUEST UNSUPPORTED: 10
NO FILE DESCRIPTOR
POLL
POLL
POLL
POLL
IOCTL REQUEST: 3
IOCTL REQUEST: 10
IOCTL REQUEST UNSUPPORTED: 10
NO FILE DESCRIPTOR
POLL
POLL
POLL
POLL
IOCTL REQUEST: 3
IOCTL REQUEST: 10
IOCTL REQUEST UNSUPPORTED: 10
NO FILE DESCRIPTOR
POLL
POLL
POLL
POLL
IOCTL REQUEST: 3
IOCTL REQUEST: 4
IOCTL REQUEST TO SOCKET
IOCTL REQUEST SOCK: 4
2023-08-12 17:43:35 [darwin@macos] [INFO] [microdot.service] GET /temp 200
# HTTP REQUEST
POLL
POLL
POLL
POLL
POLL
POLL
POLL
POLL
POLL
2023-08-12 17:46:32 [darwin@macos] [INFO] [stats.service] GET /debug HTTP/1.1
2023-08-12 17:46:32 [darwin@macos] [INFO] [stats.service] HTTP/1.1 200 OK
POLL
....
POLL # (NON STOP UNTIL NEX HTTPS REQUEST)
# HTTPS REQUEST
Request line
IOCTL REQUEST: 10
IOCTL REQUEST UNSUPPORTED: 10
NO FILE DESCRIPTOR # (STUCK HERE UNTIL NEXT HTTP REQUEST)

```
So it's definitely related to #12138 and I have no idea how to fix this but I hope these hints can help 👍🏼

Reply:
I think I've found what's wrong and my conclusion is that `poll_set_all_are_fds(poll_set)` in `extmod/modselect.c` does not work as expected. I've change that for this function 
```C
+// For each object in the poll set check if it has a file descriptor
+//
+#if MICROPY_PY_SELECT_POSIX_OPTIMISATIONS
+STATIC bool poll_set_check_all_fd(poll_set_t *poll_set, size_t *rwx_num) {
+
+    for (mp_uint_t i = 0; i < poll_set->map.alloc; ++i) {
+        if (!mp_map_slot_is_filled(&poll_set->map, i)) {
+            continue;
+        }
+
+        poll_obj_t *poll_obj = MP_OBJ_TO_PTR(poll_set->map.table[i].value);
+
+
+        if (poll_obj->pollfd == NULL) {
+            // Object does not have a file descriptor so will be polled separately by poll_set_poll_once().
+            return false;
+        }
+    }
+    return true;
+}
+#endif
+
```
Which checks if all poll objects in the poll set have a file descriptor or not. This fix the mentioned bug above 👍🏼 
I could do a separate PR for this if necessary. 👀

Reply:
Yes I see the problem.  Good work finding it!

But I think it should be solved a different way.  I'll push a fix with a test.

Reply:
FWIW I did try to solve it fixing the `poll_set_all_are_fds` logic checking the `poll_set` length but that turned out to be a slippery slope... 😕 so this was the only reliable method I could think of instead...

Reply:
See #12224 for a fix.

Reply:
That PR is now merged to master.

Reply:
Rebased and all CI tests passing, I guess this is finally ready!